File: bootstrap/src/main/java/net/dontdrinkandroot/wicket/bootstrap/css/Spacing.java
package net.dontdrinkandroot.wicket.bootstrap.css;

import net.dontdrinkandroot.wicket.css.CssClass;

/**
 * @author Philip Washington Sorst <philip@sorst.net>
 */
public class Spacing implements CssClass
{
    private Property property;

    private Side side;

    private Size size;

    public Spacing(Property property, Size size)
    {
        this.property = property;
        this.size = size;
    }

    public Spacing(Property property, Side side, Size size)
    {
        this.property = property;
        this.side = side;
        this.size = size;
    }

    @Override
    public String getClassString()
    {
        StringBuilder classString = new StringBuilder(this.property.getValue());
        if (null != this.side) {
            classString.append(this.side.getValue());
        }
        classString.append("-");
        classString.append(this.size.getValue());

        return classString.toString();
    }

    public enum Property
    {
        MARGIN("m"),
        PADDING("");

        private final String value;

        Property(String value)
        {
            this.value = value;
        }

        public String getValue()
        {
            return this.value;
        }
    }

    public enum Side
    {
        TOP("t"),
        BOTTOM("b"),
        LEFT("l"),
        RIGHT("r"),
        X("x"),
        Y("y");

        private final String value;

        Side(String value)
        {
            this.value = value;
        }

        public String getValue()
        {
            return this.value;
        }
    }

    public enum Size
    {
        ZERO("0"),
        QUARTER("1"),
        HALF("2"),
        FULL("3"),
        ONEHALF("4"),
        THREE("3"),
        AUTO("auto");

        private final String value;

        Size(String value)
        {
            this.value = value;
        }

        public String getValue()
        {
            return this.value;
        }
    }
}

File: bootstrap/src/main/kotlin/net/dontdrinkandroot/wicket/bootstrap/css/Spacing.kt
package net.dontdrinkandroot.wicket.bootstrap.css

import net.dontdrinkandroot.wicket.css.CssClass

class Spacing(
    private var property: Property,
    private var size: Size,
    private var sides: Side? = null,
    private var breakpoint: Breakpoint? = null
) : CssClass {

    override val classString: String
        get() {
            val classString = StringBuilder(property.value)
            sides?.let { classString.append(it.value) }
            breakpoint?.let {
                classString.append("-")
                classString.append(it.classString)
            }
            classString.append("-")
            classString.append(size.value)
            return classString.toString()
        }

    enum class Property(val value: String) {
        MARGIN("m"),
        PADDING("p");
    }

    enum class Side(val value: String) {
        TOP("t"),
        BOTTOM("b"),
        START("s"),
        END("e"),
        X("x"),
        Y("y");
    }

    enum class Size(val value: String) {
        NONE("0"),
        QUARTER("1"),
        HALF("2"),
        FULL("3"),
        ONEHALF("4"),
        THRICE("5"),
        AUTO("auto");
    }

    companion object {

        val MARGIN_BOTTOM_HALF = Spacing(Property.MARGIN, Size.HALF, Side.BOTTOM)
        val MARGIN_BOTTOM_FULL = Spacing(Property.MARGIN, Size.FULL, Side.BOTTOM)
        val MARGIN_END_AUTO = Spacing(Property.MARGIN, Size.AUTO, Side.END)
    }
}